homeassistant/components/camera/__init__.py

```diff
@@ -284,6 +287,51 @@ def _get_camera_from_entity_id(hass: HomeAssistant, entity_id: str) -> Camera:
     return cast(Camera, camera)
 
 
+def async_register_rtsp_to_web_rtc_provider(
+    hass: HomeAssistant,
+    domain: str,
+    provider: Callable[[str, str], Awaitable[str | None]],
+) -> Callable[[], None]:
+    """Register an RTSP to WebRTC provider.
+
+    Integrations may register a Callable that accepts a `stream_source` and
+    SDP `offer` as an input, and the output is the SDP `answer`. An implementation
+    may return None if the source or offer is not eligible or throw HomeAssistantError
+    on failure. The first provider to satisfy the offer will be used.
+    """
+    if DOMAIN not in hass.data:
+        raise ValueError("Unexpected state, camera not loaded")
+
+    def remove_provider() -> None:
+        if domain in hass.data[DATA_RTSP_TO_WEB_RTC]:
+            del hass.data[DATA_RTSP_TO_WEB_RTC]
+        hass.async_create_task(_async_refresh_providers(hass))
+
+    hass.data.setdefault(DATA_RTSP_TO_WEB_RTC, {})
+    hass.data[DATA_RTSP_TO_WEB_RTC][domain] = provider
+    hass.async_create_task(_async_refresh_providers(hass))
+    return remove_provider
+
+
+async def _async_refresh_providers(hass: HomeAssistant) -> None:
+    """Check all cameras for any state changes for registered providers."""
+    component: EntityComponent = hass.data[DOMAIN]
+    for entity in component.entities:
+        camera = cast(Camera, entity)
+        if await camera.async_refresh_providers():
+            camera.async_write_ha_state()
```

Comment:
Should we do `asyncio.gather` here?

Reply:
Yes, getting the stream source may send RPCs, so that is a good suggestion.  Done.